### PR TITLE
Minigame phase

### PR DESCRIPTION
**Pull request by AznCelebration**

### CODE REVIEW

**Review by Pukami — APPROVED** [review]
Looks good, but there's a lot of phase changing stuff that I would prefer someone else to look over to make sure there are no problems with it.

File: src/game/phase/minigame_phase.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+void MinigamePhase::c_update() {
+}
+
+void MinigamePhase::s_teardown() {
+    cancel_clock_every();
+    curr_mg->s_teardown();
+
+    /*
```
Comment (metakirby5) [inline src/game/phase/minigame_phase.cpp]: delet

File: src/game_objects/minigame/empty_collider.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    events::add_rigidbody_event(rigid);
+
+    // Set position
+    set_position(glm::vec3(0,-3,0));
+
+    notify_on_collision = true;
+}
+
+void EmptyCollider::s_on_collision(GameObject *other) {
+    if (other->tag == "PENGUIN") {
```
Comment (metakirby5) [inline src/game_objects/minigame/empty_collider.cpp]: Can we use `dynamic_cast` here instead? Using strings isn't refactor-safe.

File: src/game_objects/minigame/penguin.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
+#include "penguin.h"
+#include "asset_loader.h"
+#include "events.h"
+
+Penguin::Penguin(int id) : GameObject(id) {
+    tag = "PENGUIN";
+
+    if (id == ON_SERVER) {
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.cpp:8]: Shouldn't move speed and spawn location be the responsibility of the spawner (e.g. `Scene`)? We can just send them in as parameters to the constructors.

File: src/game_objects/minigame/penguin.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    btVector3 to_set = t.getOrigin();
+
+    // If asserts fail, please inform Kavin
+    assert(!std::isnan(to_set.getX()));
+    assert(!std::isnan(to_set.getZ()));
+
+    transform.set_position(glm::vec3((float)to_set.getX(), (float)to_set.getY(),
+        (float)to_set.getZ()));
+
+    // Forever moving left
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.cpp]: Should this be parameterized as well?

File: src/game_objects/minigame/penguin.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    });
+}
+
+void Penguin::c_update_state(glm::mat4 mat, bool enab) {
+    anim_player.update();
+    GameObject::c_update_state(mat, enab);
+}
+
+void Penguin::reset_position() {
+    set_position(glm::vec3(10, 1, rand() % 10 - 5));
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.cpp:45]: Parameterize and switch responsibility to caller?

File: src/game_objects/minigame/penguin.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+void Penguin::setup_model() {
+    attach_model(AssetLoader::get_model("slime_blue"));
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.cpp:49]: Hey, that's not a penguin

File: src/game_objects/minigame/penguin.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+*/
+
+#include "../game_object.h"
+#include "btBulletDynamicsCommon.h"
+#include "events.h"
+
+class Penguin : public GameObject {
+private:
+    float move_speed;
+    //int row;
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.h]: What's this used for?

File: src/physics.h [hunk 8]
```diff
@@ -12,11 +12,11 @@ class Physics {
 private:
     Scene *scene;
 
-    //Bullet varaibles
+    /*//Bullet varaibles
```
Comment (metakirby5) [inline src/physics.h]: Delete if not needed?

File: src/scene/minigame_scenes/mgscene_penguin.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+void MGScenePenguin::c_update() {
+    Scene::c_update();
+}
+
+void MGScenePenguin::s_setup() {
+    // Set up penguins. Initially disabled.
+    for (int i = 0; i < MAX_PENGUINS; i++) {
+        Penguin* to_add = new Penguin();
```
Comment (metakirby5) [inline src/scene/minigame_scenes/mgscene_penguin.cpp:20]: For future reference, please use smart pointers (`std::unique_ptr` and `std::shared_ptr`) so we don't have to worry about freeing memory.

File: src/scene/minigame_scenes/mgscene_penguin.cpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        to_add->disable();
+        penguins.push_back(to_add);
+        objs.push_back(to_add);
+    }
+
+    // Set up the platform
+    GameObject* platform = new GameObject(-1);
```
Comment (metakirby5) [inline src/scene/minigame_scenes/mgscene_penguin.cpp]: Can we make this its own object?

File: src/scene/minigame_scenes/mgscene_penguin.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+void MGScenePenguin::c_update() {
+    Scene::c_update();
+}
+
+void MGScenePenguin::s_setup() {
+    // Set up penguins. Initially disabled.
+    for (int i = 0; i < MAX_PENGUINS; i++) {
+        Penguin* to_add = new Penguin();
+        to_add->disable();
```
Comment (metakirby5) [inline src/scene/minigame_scenes/mgscene_penguin.cpp]: Considering just disabling in the constructor.

**Review by metakirby5 — COMMENTED** [review]
seems ok

File: src/game_objects/minigame/penguin.cpp [hunk 12]
```diff
@@ -74,4 +53,30 @@ void Penguin::setup_model() {
...
+            btVector3 vel = rigidbody->getLinearVelocity();
+            float to_move = min(move_speed, vel.getX());
+            rigidbody->setLinearVelocity(btVector3(to_move, vel.getY(), 0));
+
+            events::dungeon::request_raycast_event(
+                transform.get_position(), transform.get_forward(), 0.6f,
+                [&](GameObject *bt_hit) {
+
+                // If there's another penguin in front of it, give it a push
+                if (bt_hit && bt_hit->tag == "PENGUIN") {
```
Comment (metakirby5) [inline src/game_objects/minigame/penguin.cpp]: Use `dynamic_cast`

File: src/scene/minigame_scenes/mgscene_penguin.cpp [hunk 13]
```diff
@@ -17,19 +18,24 @@ void MGScenePenguin::s_setup() {
     // Set up penguins. Initially disabled.
     for (int i = 0; i < MAX_PENGUINS; i++) {
         Penguin* to_add = new Penguin();
+
+        // Set spawn position and rotation
+        to_add->set_position(glm::vec3(10, 1, rand() % 10 - 5));
```
Comment (metakirby5) [inline src/scene/minigame_scenes/mgscene_penguin.cpp]: Probably a bit cleaner to just add these as constructor parameters, but meh whatever

File: src/scene/minigame_scenes/mgscene_penguin.cpp [hunk 14]
```diff
@@ -48,8 +54,8 @@ void MGScenePenguin::c_setup() {
     }
 
     // Set up the platform
-    GameObject* platform = new GameObject(-1);
-    platform->attach_model(AssetLoader::get_model("cube"));
+    Platform* platform = new Platform();
+    
     platform->initial_transform.set_scale(glm::vec3(10, 1, 5));
```
Comment (metakirby5) [inline src/scene/minigame_scenes/mgscene_penguin.cpp]: Add these to the `Platform` constructor?

**Review by metakirby5 — APPROVED** [review]
few more things

**Pull request merged**